File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_10/ControlFlowRetryTest.java
package com.fResult.reactor.ch5_10;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;
import reactor.test.StepVerifier;

@Log4j2
class ControlFlowRetryTest {
  private static final String MESSAGE = "Hello, world!";

  @Test
  void retry() {
    val errored = new AtomicBoolean(false);
    val producer = Flux.create(simulatedErrorThenEmitHello(errored));

    val retryOnError =
        producer
            .doOnError(error -> log.error("Error occurred: {}", error.getMessage()))
            .retry(1)
            .doOnNext(value -> log.info("Received value: {}", value))
            .doOnComplete(() -> log.info("Stream completed successfully"));

    StepVerifier.create(retryOnError).expectNext(MESSAGE).verifyComplete();
  }

  private Consumer<? super FluxSink<String>> simulatedErrorThenEmitHello(AtomicBoolean errored) {
    return stringSink -> {
      val isFirstTime = errored.compareAndSet(false, true);

      if (isFirstTime) {
        log.error("Emitting error: {}", RuntimeException.class.getName());
        stringSink.error(new RuntimeException("Simulated error"));
      } else {
        log.info("Emitting: {}", MESSAGE);
        stringSink.next(MESSAGE);
        stringSink.complete();
      }
    };
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_10/ControlFlowRetryTest.kt
package com.fResult.reactor.ch5_10

import org.apache.logging.log4j.LogManager
import reactor.core.publisher.Flux
import reactor.core.publisher.FluxSink
import reactor.test.StepVerifier
import kotlin.reflect.jvm.jvmName
import kotlin.test.Test

class ControlFlowRetryTest {
  private val log = LogManager.getLogger(ControlFlowRetryTest::class.java)

  @Test
  fun retry() {
    val producer = Flux.create(simulatedError2TimesThenEmitHello())
    val retryOnError = producer.retry()

    StepVerifier.create(retryOnError).expectNext("Hello, World!").verifyComplete()
  }

  private fun simulatedError2TimesThenEmitHello(): (FluxSink<String>) -> Unit {
    var tryingCounter = 0

    return { stringSink ->
      val lessThan2 = ++tryingCounter <= 2

      if (lessThan2) {
        log.error("Emitting error: {} {} time(s)", RuntimeException::class.jvmName, tryingCounter)
        stringSink.error(RuntimeException("Simulated error: $tryingCounter time(s)"))
      } else {
        val message = "Hello, World!"
        log.info("Emitting {}", message)
        stringSink.next(message)
        stringSink.complete()
      }
    }
  }
}
